packages/ra-ui-materialui/src/button/DeleteWithConfirmButton.tsx

```diff
@@ -54,27 +54,84 @@ export const DeleteWithConfirmButton = <RecordType extends RaRecord = any>(
     const translate = useTranslate();
     const record = useRecordContext(props);
     const resource = useResourceContext(props);
+    const notify = useNotify();
+    const unselect = useUnselect(resource);
+    const redirect = useRedirect();
+    const [open, setOpen] = React.useState(false);
     if (!resource) {
         throw new Error(
             '<DeleteWithConfirmButton> components should be used inside a <Resource> component or provided with a resource prop. (The <Resource> component set the resource prop for all its children).'
         );
     }
 
-    const {
-        open,
-        isPending,
-        handleDialogOpen,
-        handleDialogClose,
-        handleDelete,
-    } = useDeleteWithConfirmController({
+    const { onSuccess, onError, ...otherMutationOptions } =
+        mutationOptions || {};
+
+    const { isPending, handleDelete } = useDeleteController({
         record,
-        redirect,
+        redirect: redirectTo,
         mutationMode,
-        onClick,
-        mutationOptions,
+        mutationOptions: {
+            ...otherMutationOptions,
+            onSuccess: (data, variables, context) => {
+                setOpen(false);
+                if (onSuccess) {
+                    onSuccess(data, variables, context);
+                } else {
+                    notify(
+                        successMessage ??
+                            `resources.${resource}.notifications.deleted`,
+                        {
+                            type: 'info',
+                            messageArgs: {
+                                smart_count: 1,
+                                _: translate('ra.notification.deleted', {
+                                    smart_count: 1,
+                                }),
+                            },
+                            undoable: mutationMode === 'undoable',
+                        }
+                    );
+                    record && unselect([record.id]);
+                    redirect(redirectTo, resource);
+                }
+            },
```

Comment:
Not sure about the choice not to use `useDeleteWithConfirmController` here. It looks like you had to move much code from ra-core over to ra-ui-materialui, which seems to go the opposite way of the ongoing efforts to enhance the headless core of RA.

To me things like handling notifications, side-effects, record selection state and redirection all belong to ra-core...

Reply:
This is because we need to close the dialog at L77. We decided to do it like it with @fzaninotto because we don't want UI concerns to be handled in the base controller. As we also don't want to have default side effects in the base controller that would call user provided side effects, we must duplicate them here.

Reply:
As discussed, we'll keep the duplication